tests/test_example_meshes.cpp

```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
     bool is_oriented = false;
     std::array<long, 3> simplex_counts = std::array<long,3>{{-1,-1,-1}};
 };
+
+std::array<long, 3> trimesh_simplex_counts(const DEBUG_TriMesh& m)
+{
+    return std::array<long, 3>{{long(m.get_all(PrimitiveType::Vertex).size()),
+                                long(m.get_all(PrimitiveType::Edge).size()),
+                                long(m.get_all(PrimitiveType::Face).size())}};
+}
+
+
+int trimesh_simply_connected_components(const DEBUG_TriMesh& m, std::vector<int> &component_ids)
+{
+    component_ids.resize(m.capacity(PrimitiveType::Face), -1);
+    auto all_face_tuples = m.get_all(PrimitiveType::Face);
+    int component_id = 0;
+    // BFS
+    for (size_t i = 0; i < all_face_tuples.size(); ++i) {
+        long fid = m.id(all_face_tuples[i], PrimitiveType::Face);
+        if (component_ids[fid] != -1) {
+            continue;
+        } // visited
+
+        std::queue<long> q;
+        q.push(fid);
+        while (!q.empty()) {
+            long cur_fid = q.front();
+            q.pop();
+
+            component_ids[cur_fid] = component_id;
+            auto f_tuple = m.tuple_from_face_id(cur_fid);
+            // push all adjacent faces
+            for (int j = 0; j < 3; ++j)
+            {
+                if (!m.is_boundary(f_tuple))
+                {   
+                    auto adj_face_tuple = m.switch_tuple(f_tuple, PrimitiveType::Face);
+                    long adj_fid = m.id(adj_face_tuple, PrimitiveType::Face);
+                    if (component_ids[adj_fid] == -1) {
+                        q.push(adj_fid);
+                    }  
+                }
+                f_tuple = m.switch_tuple(m.switch_tuple(f_tuple, PrimitiveType::Vertex), PrimitiveType::Edge);
+            } 
+        }
+        
+        component_id++;
+    }
+
+    return component_id;
+}
+
+int trimesh_simply_connected_components(const DEBUG_TriMesh& m)
+{
+    std::vector<int> component_ids;
+    return trimesh_simply_connected_components(m, component_ids);
+}
+
+int trimesh_n_bd_loops(const DEBUG_TriMesh& m)
+{
+    std::vector<int> bd_loops(m.capacity(PrimitiveType::Edge), -1);
+    auto all_edge_tuples = m.get_all(PrimitiveType::Edge);
+    int bd_loop_id = 0;
+    for (size_t i = 0; i < all_edge_tuples.size(); ++i) 
+    {
+        if (!m.is_boundary(all_edge_tuples[i])) {
+            continue;
+        }
+        long eid = m.id(all_edge_tuples[i], PrimitiveType::Edge);
+        if (bd_loops[eid] != -1) {
+            continue;
+        } // visited
+
+        Tuple cur_edge = all_edge_tuples[i];
+        long cur_eid = m.id(cur_edge, PrimitiveType::Edge);
+        // find one boundary loop
+        while (bd_loops[m.id(all_edge_tuples[i], PrimitiveType::Edge)] == -1 || cur_eid != m.id(all_edge_tuples[i], PrimitiveType::Edge)) 
```

Comment:
declare variables when values are used twice.... `m.id(all_edge_tuples[i], PrimitiveType::Edge)` doesn't need to appear more than once